en/1/06-functions.md

```diff
@@ -0,0 +1,195 @@
+---
+title: Function Declarations
+actions: ['checkAnswer', 'hints']
+material:
+  editor:
+    language: rust
+    startingCode: |
+      #![no_std]
+
+      multiversx_sc::imports!();
+      multiversx_sc::derive_imports!();
+
+      #[derive(NestedEncode, NestedDecode, TopEncode, TopDecode, TypeAbi)]
+      pub struct Zombie<M: ManagedTypeApi> {
+          name: ManagedBuffer<M>,
+          dna: u64,
+      }
+
+      #[multiversx_sc::contract]
+      pub trait ZombiesContract {
+
+        #[init]
+        fn init(&self) {
+          self.dna_digits().set(16u8);
+          self.zombies_count().set(1usize);
+        }
+
+          // start here
+
+        #[storage_mapper("dna_digits")]
+        fn dna_digits(&self) -> SingleValueMapper<u8>;
+
+        #[storage_mapper("zombies_count")]
+        fn zombies_count(&self) -> SingleValueMapper<usize>;
+
+        #[storage_mapper("zombies")]
+        fn zombies(&self, id: usize) -> SingleValueMapper<Zombie<Self::Api>>;
+      }
+    answer: >
+      #![no_std]
+
+      multiversx_sc::imports!();
+      multiversx_sc::derive_imports!();
+
+      #[derive(NestedEncode, NestedDecode, TopEncode, TopDecode, TypeAbi)]
+      pub struct Zombie<M: ManagedTypeApi> {
+          name: ManagedBuffer<M>,
+          dna: u64,
+      }
+
+      #[multiversx_sc::contract]
+      pub trait ZombiesContract {
+
+        #[init]
+        fn init(&self) {
+          self.dna_digits().set(16u8);
+          self.zombies_count().set(1usize);
+        }
+
+        fn create_zombie(&self, name: ManagedBuffer, dna: u64) {
+            self.zombies_count().update(|id| {
+              self.zombies(id).set(Zombie { name, dna });
+              *id +=1;
+            });
+        }
+
+        #[storage_mapper("dna_digits")]
+        fn dna_digits(&self) -> SingleValueMapper<u8>;
+        
+        #[storage_mapper("zombies_count")]
+        fn zombies_count(&self) -> SingleValueMapper<usize>;
+
+        #[storage_mapper("zombies")]
+        fn zombies(&self, id: &usize) -> SingleValueMapper<Zombie<Self::Api>>;
+      }
+---
+
+## Functions in Rust
+
+A function declaration in Rust looks like the following:
+
+```
+pub function eat_hamburgers(&self, amount: u32) {
+
+}
+```
+
+This is a function named `eat_hamburgers` that takes 2 parameters: a reference to `self` and a `u32`. For now the body of the function is empty. Note that we're specifying the function visibility as `public`. Without the `pub` keyword the function would be private, visible only within the entity that implements it.
+
+What is a reference type you ask?
+
+Well, there are three ways in which you can pass an argument to a Rust function:
```

Comment:
For non-Rust developers or Solidity developers, I think the following list will be very abstract and I am not sure they will concretely understand. Maybe would it be good to give a concrete example of passing by value, by reference and by mutable reference.

Also if we don't need to know all 3 ways of passing right now, I would postpone the introduction as this will reduce the learning curve.